File: blog-backend/app/routers/subscriptions.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List, Dict, Any

from app import models
from app.core import security
from app.core.database import get_db
from app.schemas.subscription import (
    CategorySubscriptionCreate, AuthorSubscriptionCreate,
    NotificationResponse, SubscriptionResponse
)
from app.services.subscription_service import SubscriptionService, NotificationService
from app.utils.pagination import PaginationParams, PagedResponse

router = APIRouter(prefix="/subscriptions", tags=["subscriptions"])

@router.post("/categories", status_code=status.HTTP_201_CREATED)
async def subscribe_to_category(
    subscription: CategorySubscriptionCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Subscribe to a category.
    """
    success = SubscriptionService.subscribe_to_category(
        db=db,
        user_id=current_user.id,
        category_id=subscription.category_id
    )

    if not success:
        return {"message": "Already subscribed to this category"}

    return {"message": "Successfully subscribed to category"}

@router.delete("/categories/{category_id}")
async def unsubscribe_from_category(
    category_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Unsubscribe from a category.
    """
    success = SubscriptionService.unsubscribe_from_category(
        db=db,
        user_id=current_user.id,
        category_id=category_id
    )

    if not success:
        return {"message": "Not subscribed to this category"}

    return {"message": "Successfully unsubscribed from category"}

@router.post("/authors", status_code=status.HTTP_201_CREATED)
async def subscribe_to_author(
    subscription: AuthorSubscriptionCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Subscribe to an author.
    """
    success = SubscriptionService.subscribe_to_author(
        db=db,
        subscriber_id=current_user.id,
        author_id=subscription.author_id
    )

    if not success:
        return {"message": "Already subscribed to this author"}

    return {"message": "Successfully subscribed to author"}

@router.delete("/authors/{author_id}")
async def unsubscribe_from_author(
    author_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Unsubscribe from an author.
    """
    success = SubscriptionService.unsubscribe_from_author(
        db=db,
        subscriber_id=current_user.id,
        author_id=author_id
    )

    if not success:
        return {"message": "Not subscribed to this author"}

    return {"message": "Successfully unsubscribed from author"}

@router.get("", response_model=SubscriptionResponse)
async def get_subscriptions(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Get current user's subscriptions.
    """
    subscriptions = SubscriptionService.get_user_subscriptions(db=db, user_id=current_user.id)

    return {
        "user_id": current_user.id,
        "category_ids": subscriptions["category_ids"],
        "author_ids": subscriptions["author_ids"]
    }

# Notifications endpoints
@router.get("/notifications", response_model=PagedResponse[NotificationResponse])
async def get_notifications(
    unread_only: bool = Query(False, description="Get only unread notifications"),
    pagination: PaginationParams = Depends(),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Get current user's notifications.
    """
    notifications = NotificationService.get_user_notifications(
        db=db,
        user_id=current_user.id,
        skip=pagination.skip,
        limit=pagination.page_size,
        unread_only=unread_only
    )

    # Get total count
    total = db.query(models.Notification).filter(
        models.Notification.user_id == current_user.id
    )

    if unread_only:
        total = total.filter(models.Notification.is_read == False)

    total = total.count()

    return PagedResponse.create(notifications, total, pagination)

@router.put("/notifications/{notification_id}/read")
async def mark_notification_as_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Mark a notification as read.
    """
    notification = NotificationService.mark_notification_as_read(
        db=db,
        notification_id=notification_id,
        user_id=current_user.id
    )

    return {"message": "Notification marked as read"}

@router.put("/notifications/read-all")
async def mark_all_notifications_as_read(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Mark all notifications as read.
    """
    count = NotificationService.mark_all_notifications_as_read(
        db=db,
        user_id=current_user.id
    )

    return {"message": f"{count} notifications marked as read"}

@router.delete("/notifications/{notification_id}")
async def delete_notification(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Delete a notification.
    """
    success = NotificationService.delete_notification(
        db=db,
        notification_id=notification_id,
        user_id=current_user.id
    )

    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )

    return {"message": "Notification deleted"}

File: blog-frontend/src/api/modules/subscriptions.js
/**
 * 订阅相关 API
 * 封装所有与订阅相关的网络请求
 */

import apiClient from '../core/apiClient';
import { API_PATHS } from '../core/apiPaths';

/**
 * 获取用户订阅
 * @returns {Promise} 返回用户订阅信息
 */
export const getUserSubscriptions = () => {
  return apiClient.get(API_PATHS.SUBSCRIPTIONS.BASE);
};

/**
 * 订阅分类
 * @param {number} categoryId - 分类ID
 * @returns {Promise} 返回订阅结果
 */
export const subscribeToCategory = (categoryId) => {
  return apiClient.post(API_PATHS.SUBSCRIPTIONS.CATEGORIES, { category_id: categoryId });
};

/**
 * 取消订阅分类
 * @param {number} categoryId - 分类ID
 * @returns {Promise} 返回取消订阅结果
 */
export const unsubscribeFromCategory = (categoryId) => {
  return apiClient.delete(`${API_PATHS.SUBSCRIPTIONS.CATEGORIES}/${categoryId}`);
};

/**
 * 订阅作者
 * @param {number} authorId - 作者ID
 * @returns {Promise} 返回订阅结果
 */
export const subscribeToAuthor = (authorId) => {
  return apiClient.post(API_PATHS.SUBSCRIPTIONS.AUTHORS, { author_id: authorId });
};

/**
 * 取消订阅作者
 * @param {number} authorId - 作者ID
 * @returns {Promise} 返回取消订阅结果
 */
export const unsubscribeFromAuthor = (authorId) => {
  return apiClient.delete(`${API_PATHS.SUBSCRIPTIONS.AUTHORS}/${authorId}`);
};

/**
 * 创建邮件订阅
 * @param {Object} data - 订阅数据
 * @param {string} data.email - 邮箱地址
 * @param {string} data.subscription_type - 订阅类型 (author, category, all)
 * @param {number|null} data.reference_id - 引用ID (作者ID或分类ID，对于all类型为null)
 * @returns {Promise} 返回订阅结果
 */
export const createEmailSubscription = (data) => {
  return apiClient.post(API_PATHS.EMAIL_SUBSCRIPTIONS.BASE, data);
};

/**
 * 通过令牌取消邮件订阅
 * @param {string} token - 订阅令牌
 * @returns {Promise} 返回取消订阅结果
 */
export const unsubscribeByToken = (token) => {
  return apiClient.post(API_PATHS.EMAIL_SUBSCRIPTIONS.UNSUBSCRIBE, { token });
};

/**
 * 通过邮箱取消订阅
 * @param {string} email - 邮箱地址
 * @param {string} [subscription_type] - 订阅类型 (可选)
 * @param {number} [reference_id] - 引用ID (可选)
 * @returns {Promise} 返回取消订阅结果
 */
export const unsubscribeByEmail = (email, subscription_type = null, reference_id = null) => {
  const params = { email };

  if (subscription_type) {
    params.subscription_type = subscription_type;
  }

  if (reference_id !== null) {
    params.reference_id = reference_id;
  }

  return apiClient.post(API_PATHS.EMAIL_SUBSCRIPTIONS.UNSUBSCRIBE_EMAIL, null, { params });
};

/**
 * 获取邮件订阅列表
 * @param {Object} params - 查询参数
 * @param {string} [params.email] - 按邮箱筛选
 * @param {string} [params.subscription_type] - 按订阅类型筛选 (author, category, all)
 * @param {number} [params.reference_id] - 按引用ID筛选
 * @param {boolean} [params.is_active] - 是否只返回活跃订阅
 * @returns {Promise} 返回订阅列表
 */
export const getEmailSubscriptions = (params = {}) => {
  return apiClient.get(API_PATHS.EMAIL_SUBSCRIPTIONS.BASE, { params });
};

/**
 * 更新邮件订阅状态
 * @param {number} id - 订阅ID
 * @param {Object} data - 更新数据
 * @param {boolean} data.is_active - 是否活跃
 * @returns {Promise} 返回更新结果
 */
export const updateEmailSubscription = (id, data) => {
  return apiClient.patch(`${API_PATHS.EMAIL_SUBSCRIPTIONS.BASE}/${id}`, data);
};

export default {
  getUserSubscriptions,
  subscribeToCategory,
  unsubscribeFromCategory,
  subscribeToAuthor,
  unsubscribeFromAuthor,
  createEmailSubscription,
  unsubscribeByToken,
  unsubscribeByEmail,
  getEmailSubscriptions,
  updateEmailSubscription
};
